### DISCUSSION TITLE
pointer types in unsafe code aren't supported

### DISCUSSION THREAD

**Discussion by snikeguo** [nanoframework/Home#1108] [Ideas]
### Tool

Visual Studio extension

### Description

c# code:
byte *ptr =(byte *)0x80000000;
byte *d=*ptr;

build failed:
pointer types in unsafe code aren't supported

**Comment by josesimoes** [comment]
Hello @snikeguo! Thanks for reaching out.
This is not a bug, it's by design. We don't have any plans to add support for these.

**Comment by snikeguo** [comment]
![捕获(1)](https://user-images.githubusercontent.com/20830435/182743889-e7d94798-c5f8-4219-a84f-1e8874980a81.jpg)
@josesimoes

**Comment by snikeguo** [comment]
I want to use c# to write drivers:
`
u32 *reg= (u32*) 0x12345678;
var oldvalue=*reg;
*reg=0x87654321;
`
Are you worried that others can read the firmware?

**Reply by josesimoes** [reply]
Allowing this opens the possibility to read/write from/to memory and even registers.
That has inpact on security and privacy not to mention that it can render a system unusable or unstable by allowing to change any content addressable in memory.

**Reply by snikeguo** [reply]
I think this function can be added, and the right to select is on the user. Just like C ++: You can not use the meta -programming function, but I can't without it!There are 20%of senior users using it.

**Reply by josesimoes** [reply]
There's more to it than just "add it". I haven't wrote that it's not possible, rather that's not on our plans to implement it. For the reasons mentioned. 
I don't have access to the same statistics that you have, obviously. 😅  But from the output of nanoFramework community, as far as I can recall, you're the first and only one to request it.

Can you elaborate on your usage scenario please?

**Comment by Tajbiul-Rawol** [comment]
Why is there no plan to support it? I mean why limit the usage of pointers and go really low and instead worry about safety? Isn't this the point to allow the flexibility to write safe code with C# but you can also shoot yourself in the foot if you want to?

**Comment by Ellerbach** [comment]
> Why is there no plan to support it?

Because of the way the stack is managed.

>  I mean why limit the usage of pointers and go really low and instead worry about safety?

For this, you can run your own C/C++ code and use interop to consume it in C#. All nanoFramework is built like this!

**Reply by Tajbiul-Rawol** [reply]
Is the stack not managed the same way as the normal .NET runtime? I thought it did. I’m sure you have reasons to limit it. But I think people in the community would appreciate if given the flexibility allowing pointer stuff in c# itself. It would be cool. I myself always wanted c# to go to titly bit level which I think nanoframework is perfect for.

**Reply by alberk8** [reply]
It is a cut down version of the big .Net runtime so there are no one to one in terms of complete features.   Question, Why not allow user to poke native pointer?.   Allowing it will be a support nightmare, user will complain that the Device is not seen in the Device Explorer when wrong pointer is accessed.  Why, because the underlying RTOS will reboot if there is an exception, not like on a full fledge OS where these exception will be caught (if poking user space memory).  Poking native pointer in NF is like poking the Kernel driver in any OS, you can expect it to blow up and not easy to debug.   

Hope this help to some what explain the issue.

**Reply by Tajbiul-Rawol** [reply]
Okay that makes sense. Then Does .NET nano-framework support NativeMemory and Span? which I think are unsafe pointer alternatives but serve the same purpose but in a safer way. Will NF allow Native AOT compilation in the future?

**Reply by alberk8** [reply]
1) NativeMemory,  I guess you are asking to access to the native side memory, if that is the case no, not from the managed side. You can write you own interop to access the well defined memory space in native memory and this involves you compiling a custom firmware and writing you managed code (class library))
2) [Span](https://docs.nanoframework.net/api/System.SpanByte.html),  we do have that but slightly modified as we do not have generics yet
3) Don't think so,  unless someone is willing to give it a try.  For argument sake, with the full force of Microsoft, NativeAOT is still not complete,  for eg. You still cannot compile a NativeAOT on a asp.net project with Controller/Blazor.

**Reply by snikeguo** [reply]
> 1. NativeMemory,  I guess you are asking to access to the native side memory, if that is the case no, not from the managed side. You can write you own interop to access the well defined memory space in native memory and this involves you compiling a custom firmware and writing you managed code (class library))
> 2. [Span](https://docs.nanoframework.net/api/System.SpanByte.html),  we do have that but slightly modified as we do not have generics yet
> 3. Don't think so,  unless someone is willing to give it a try.  For argument sake, with the full force of Microsoft, NativeAOT is still not complete,  for eg. You still cannot compile a NativeAOT on a asp.net project with Controller/Blazor.

You don't need to expect the nanoFramework anymore. We are porting coreCLR to STM32: NuttX (providing pthread, GLib, FileSystem,heap memory manager) + coreCLR (with JIT support). Therefore, the hardware requires an MCU + SDRAM + external Flash (NOR Flash/NAND flash/TF card).
We've already successfully run "hello world" in JIT mode with garbage collection (GC). I'm currently awaiting the company's decision on whether to open-source it.

**Reply by Tajbiul-Rawol** [reply]
incredible, In that case you can do manual memory allocation, native aot and use the NativeMemory class? if this is open sourced will it come with a build pipeline or just as a framework?

**Reply by snikeguo** [reply]
> incredible, In that case you can do manual memory allocation, native aot and use the NativeMemory class? if this is open sourced will it come with a build pipeline or just as a framework?

NO AOT，ONLY  JIT。 Currently, only the HelloWorld functionality has been implemented( contain System.Core.Private.dll), and no other drivers have been developed. In fact, it is possible to use C# to develop TCP/IP, CAN, IIC, and other drivers.

**Reply by snikeguo** [reply]
> incredible, In that case you can do manual memory allocation, native aot and use the NativeMemory class? if this is open sourced will it come with a build pipeline or just as a framework?

Currently, I'm also waiting for the .NET team to port the interpreter.

**Reply by Tajbiul-Rawol** [reply]
Hopefully your team decides to make it open Source.

**Reply by alberk8** [reply]
May I know what is the minimum RAM requirement to run CoreCLR ?

**Reply by snikeguo** [reply]
> May I know what is the minimum RAM requirement to run CoreCLR ?

1MB with JIT
